iothub_client/inc/iothub_client_properties.h

```diff
@@ -0,0 +1,298 @@
+*
+*           This API does not perform any network I/O.  It only translates IOTHUB_CLIENT_WRITABLE_PROPERTY_RESPONSE into
+*           a byte array.  Applications should use APIs such as IoTHubDeviceClient_LL_SendPropertiesAsync to send the 
+*           serialized data.
+*
+* @return   IOTHUB_CLIENT_OK upon success or an error code upon failure.
+*/
+IOTHUB_CLIENT_RESULT IoTHubClient_Serialize_WritablePropertyResponse(
+    const IOTHUB_CLIENT_WRITABLE_PROPERTY_RESPONSE* properties,
+    size_t numProperties,
+    const char* componentName,
+    unsigned char** serializedProperties,
+    size_t* serializedPropertiesLength);
+
+/**
+* @brief   Frees serialized properties that were initially allocated with IoTHubClient_Serialize_ReportedProperties 
+*          or IoTHubClient_Serialize_WritablePropertyResponse.
+*
+* @param[in]  serializedProperties Properties to free.
+*/
+void IoTHubClient_Serialize_Properties_Destroy(unsigned char* serializedProperties);
+
+typedef struct IOTHUB_CLIENT_PROPERTY_ITERATOR_TAG* IOTHUB_CLIENT_PROPERTY_ITERATOR_HANDLE;
+
+/**
+* @brief   Setup an iterator to enumerate through Plug and Play properties.
+*
+* @param[in]  payloadType              Whether the payload is a full set of properties or only a set of updated 
+*                                      writable properties.
+* @param[in]  payload                  Payload containing properties from Azure IoT that is to be deserialized. 
+* @param[in]  payloadLength            Length of @p payload.
+* @param[in]  componentsInModel        Optional array of components that correspond to the DTDLv2 model.  Can be NULL 
+*                                      for models that don't contain sub-components.
+* @param[in]  numComponentsInModel     Number of entries in @p componentsInModel.
+* @param[out] propertyIteratorHandle   Returned handle used for subsequent iteration calls.
+* @param[out] propertiesVersion        Returned version of the writable properties.  This is used when acknowledging a 
+*                                      write property update.
+* 
+* @remarks  Applications typically will invoke this API in their IOTHUB_CLIENT_PROPERTIES_RECEIVED_CALLBACK 
+*           implementation.  Many of the parameters this function takes should come directly from the 
+*           IOTHUB_CLIENT_PROPERTIES_RECEIVED_CALLBACK itself.
+*
+*           Applications must call @p IoTHubClient_Deserialize_Properties_DestroyIterator to free @p 
+*           propertyIteratorHandle on completion.
+*
+* @return   IOTHUB_CLIENT_OK upon success or an error code upon failure.
+*/
+IOTHUB_CLIENT_RESULT IoTHubClient_Deserialize_Properties_CreateIterator(
+    IOTHUB_CLIENT_PROPERTY_PAYLOAD_TYPE payloadType,
+    const unsigned char* payload,
+    size_t payloadLength,
+    const char** componentsInModel,
+    size_t numComponentsInModel,
+    IOTHUB_CLIENT_PROPERTY_ITERATOR_HANDLE* propertyIteratorHandle,
+    int* propertiesVersion);
```

Comment:
>    int* propertiesVersion [](http://example.com/codeflow?start=0&length=26)

this API only return write property, but document also can contain reported property version. Should we also expose reported version as argument?

Or

we can have these version cached in the handle when we create them, and allow user to access these version using Get Api's?